src/main/scala/com/mongodb/spark/connection/MongoClientCache.scala

```diff
@@ -18,76 +18,67 @@
 //scalastyle:on
 package com.mongodb.spark.connection
 
+import java.util.concurrent.atomic.AtomicReference
 import java.util.concurrent.{Executors, ThreadFactory, TimeUnit}
 
+import com.mongodb.MongoClient
+import com.mongodb.spark.{Logging, MongoClientFactory}
+
 import scala.annotation.tailrec
 import scala.collection.JavaConverters._
-import scala.collection.concurrent.TrieMap
 import scala.concurrent.duration.Duration
 import scala.util.{Failure, Success, Try}
 
-import com.mongodb.spark.Logging
-
-import com.mongodb.spark.MongoClientFactory
-import com.mongodb.{MongoClient, ServerAddress}
-
 /**
- * A lockless cache for MongoClients.
+ * A lockless cache for a MongoClient.
  *
- * Allows multiple users access to MongoClients. Closes a `MongoClient` when they're are no longer used.
+ * Allows multiple users access to a MongoClient. Closes the `MongoClient` when they're are no longer used.
  *
- * @param keepAlive the duration to keep alive any given MongoClient so that it can be reused by another consumer
+ * @param keepAlive the duration to keep alive a MongoClient so that it can be reused by another consumer
  */
 private[spark] final class MongoClientCache(keepAlive: Duration) extends Logging {
 
   private val refCounter = new MongoClientRefCounter
-  private val cache = new TrieMap[Option[ServerAddress], MongoClient]
-  private val clientToKey = new TrieMap[MongoClient, Option[ServerAddress]]
-  private val deferredReleases = new TrieMap[MongoClient, ReleaseTask]
+  private val cache = new AtomicReference[MongoClient]()
+  private val deferredReleases = new AtomicReference[ReleaseTask]
 
-  // scalastyle:off cyclomatic.complexity
   @tailrec
-  def acquire(key: Option[ServerAddress], mongoClientFactory: MongoClientFactory): MongoClient = {
-    cache.get(key) match {
+  def acquire(mongoClientFactory: MongoClientFactory): MongoClient = {
```

Comment:
Actually, it can't be simplified like I have, as input and output uri's could be different so we'd need multiple clients if they were.  Reverting the change.